### ISSUE TITLE
Consider adding addAllFoo(Stream<T>) to Builder

### ISSUE THREAD

**Issue by bhamiltoncx** [immutables/immutables#593]
Java 8 feature request:

`Builder`s currently support `addAllFoo(Iterable<T>)`, but they don't support Java 8 `Stream<T>`, so we have to do an extra copy each time we add items from a `Stream` to a `Builder`.

It'd be nice if we could introduce `addAllFoo(Stream<T>)` and friends to the `Builder`s.

**Comment by elucash** [comment]
Good idea! Definitely, there will be a lot less noise with collectors. Now, any built-in container support use only pre-java 8 APIs, need to think if there will be any problems with detecting java8 and generating such methods. I understand that proposing custom encodings for this would be a sad idea as it would be sweet to have it for all built-in recognizable containers.

**Comment by ldriscoll** [comment]
I'd be a little cautious of this, since a Stream can only be consumed once, perhaps implementing a collector would be better.

**Comment by bhamiltoncx** [comment]
Luke is right, a method for each field which returns a Collector is more
appropriate.

**Comment by bhamiltoncx** [comment]
Java 7 has been EOL since April 2015. Should we consider requiring Java 8?

Lambdas would probably make the codebase prettier.

**Comment by bhamiltoncx** [comment]
(I don't know the requirements of your big clients, but Buck at least requires Java 8. :)

**Comment by elucash** [comment]
there a number of people using Immutables for Java7 and even Java 6 (with some feature loss)

the other thing is I don't get how collectors are any better than consuming streams. Streams will be exhausted in either case

```java
Consumer<Stream<T>> consumer = builder::addAllFoo
Collector<T....> collector = builde.fooCollector();
```
in their side effects, `consumer` and `collector`  are the same

```java
builder.addAllFoo(collection.stream().map(x -> s)));  // may be more compelling than

collection.stream().map(x -> s).collect(builder.fooCollector());
```

**Comment by bhamiltoncx** [comment]
I think the idea is you can nest/group `Collector`s in various interesting ways, but once you pass a `Stream<T>` to something, it's gone.

**Comment by bhamiltoncx** [comment]
(However, I don't know if that really matters much for Builders.)

**Comment by elucash** [comment]
@ldriscoll @bhamiltoncx  any good examples with collectors?

**Comment by bhamiltoncx** [comment]
This might be a stretch, but one interesting example is parallel collecting:

https://docs.oracle.com/javase/tutorial/collections/streams/parallelism.html#concurrent_reduction

**Comment by elucash** [comment]
certainly, a good example of collectors, but how we would use this in conjunction with Immutabes' builder collectors, how we could compose them?

**Comment by bhamiltoncx** [comment]
Yeah, I don't think there would be much benefit to composing Immutables' `Builder` collectors.

The main thing we have to think about is: how do we plan on extracting items from the `Stream<T>`? I think there's only two ways (via a `Collector` and via `Stream.toArray()`).

I think the Java 8 runtime is smart enough to prefetch and batch items from the `Stream<T>` to improve throughput when passing to a `Collector`, but it can't do that if we manually extract items from the `Stream` (e.g. with `Stream.toArray()`.)

**Comment by elucash** [comment]
I mostly care about API here. In the implementation, we would use some sort of collectors or just forEach to add things to the internal collection builders.

**Comment by bhamiltoncx** [comment]
Yeah, I can see arguments for both styles of API that you mentioned. I like them both. :)

**Comment by elucash** [comment]
just to note:

```java
collection.stream().map(x -> x....).forEach(builder::addFoo);
// vs
collection.stream().map(x -> x....).collect(builder.fooCollector());
```

**Comment by bhamiltoncx** [comment]
Right. I think the forEach form must be serial, but the collect form may be
parallel.

**Comment by elucash** [comment]
builders are not thread-safe, so parallel collectors are in question.

**Comment by bhamiltoncx** [comment]
> builders are not thread-safe

Not yet, anyway! ;)
